### ISSUE TITLE
Allow Jobs to "output" a File (ie Download to Browser)

### ISSUE THREAD

**Issue by lampwins** [nautobot/nautobot#3352]
As a Nelly the Network Automation Engineer I want to write a job which munges data and produces a file that I can download from the UI when the job completes so that perform some unnatural UX operation to retrieve that output (e.g. obtain the file from some other location, copy paste output text, etc).

**Comment by Kircheneer** [comment]
I have something that somewhat works, just putting this here for visibility. Not recommending the use of this.

```python
from django.urls import reverse

...

# Put the following stuff into a Jobs `run` method
def run(...):
  ...
  file_name = f"{self.job_result.pk}.txt"
  encoding = "utf-8"
  with io.BytesIO(b"text_to_put_into_file") as buffer:
      # Upload file to Nautobot's database
      in_memory_uploaded_file = InMemoryUploadedFile(
          file=buffer, field_name=None, name=file_name, content_type="text/csv", size=None, charset=encoding
      )
      file_proxy_pk = self.save_file(in_memory_uploaded_file)
  
  # Make file available for download
  file_name = FileProxy.objects.get(pk=file_proxy_pk).file.name
  self.log_info(
      obj=None,
      message=f"Download CSV [here]({reverse('db_file_storage.download_file')}?name={file_name})."
      f" It is recommended that you delete this file"
      f" [here]({reverse('admin:extras_fileproxy_change', args=[file_proxy_pk])}) once you have downloaded"
      f" it.",
  )
  ```

**Comment by bryanculver** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Nautobot-Core-Product-64a8783b7a3f4b1225973a0c/issues/nautobot/nautobot/3352?planning-poker) with Zenhub @glennmatthews @gsnider2195 @HanlinMiao @jathanism @timizuoebideri1

**Comment by bryanculver** [comment]
We should document what the file size limit is with `FileProxy` since at present it is using the database as a "file system".

**Comment by jathanism** [comment]
> We should document what the file size limit is with `FileProxy` since at present it is using the database as a "file system".

I think we had discussed this but it's not captured here, but seeing as the underlying model for `FileProxy.file` is `FileAttachment.bytes`, which is a `BinaryField` without a `max_length`, the file size will be limited by the database backend.

- For MySQL, a `BinaryField` is represented as a [`LONGBLOB`](https://dev.mysql.com/doc/refman/8.0/en/storage-requirements.html), which has a maximum size of 4GB.
- For PostgreSQL, a `BinaryField` is represented as a [`bytea`](https://stackoverflow.com/a/58951936), which has a maximum size of 1GB.

Keep in mind these are theoretical limits and allowing files of such sizes would likely be an incredibly bad idea causing service to halt. We might consider actually putting a hard limit on the database field itself by setting `FileAttachment.bytes.max_length` in a schema migration.

**Comment by smk4664** [comment]
I have this created in one environment, but it is not using the FIleProxy model. It is using a model like the Image Attachment model, but supports multiple file types. I can do a PoC of this, if y'all would like. Ideally, instead of storing the file to the database, it would be stored in Django Storages.

**Comment by glennmatthews** [comment]
Implemented for 2.1 by #4723
